File: J9Ch/src/J_9_Dynamic_Programming/Optional_5/_117_Jump_Game_II_Medium.cpp
/** 117 Jump Game II */


class Solution {
public:
    int jump(vector<int> A) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        int n = A.size();
        int minstep = 0;
        int ldist = 0;
        int hdist = 0;
        if (n == 1) return 0;
        while(ldist <= hdist) {
            ++minstep;
            int lasthdist = hdist;
            for(int i = ldist; i <= lasthdist; ++i) {
                int possibledist = i + A[i];
                if (possibledist >= n-1)
                    return minstep;
                if (possibledist > hdist) {
                    hdist = possibledist;
                }
            }
            ldist = lasthdist + 1;
        }
        return 0;
    }
};

File: J9Ch/src/J_9_DP/Optional_5/python_cpp/_117_Jump_Game_II_Medium.py
# /** 117 Jump Game II


class Solution:
    # @param A, a list of integers
    # @return an integer
    def jump(self, A):
        p = [0]
        for i in range(len(A) - 1):
            while(i + A[i] >= len(p) and len(p) < len(A)):
                p.append(p[i] + 1)
        return p[-1]
